cpp/src/dictionary/detail/concatenate.cu

```diff
@@ -0,0 +1,268 @@
+    : columns_ptrs{columns.size()}
+  {
+    std::transform(
+      columns.begin(), columns.end(), columns_ptrs.begin(), [](auto& cv) { return &cv; });
+  }
+
+  /**
+   * @brief Return the first keys().type of the dictionary columns.
+   */
+  data_type get_keys_type()
+  {
+    dictionary_column_view dict_view(**std::find_if(
+      columns_ptrs.begin(), columns_ptrs.end(), [](auto pcv) { return pcv->size() > 0; }));
+    return dict_view.keys().type();
+  }
+
+  /**
+   * @brief Create the offsets pair for the concatenated columns.
+   *
+   * Both vectors have the length of the number of input columns.
+   * The sizes of each child (keys and indices) of the individual columns
+   * are used to create the offsets.
+   *
+   * @param stream Stream used for allocating the output rmm::device_uvector.
+   * @return Vector of offsets_pair objects for keys and indices.
+   */
+  rmm::device_uvector<offsets_pair> create_children_offsets(cudaStream_t stream)
+  {
+    std::vector<offsets_pair> offsets(columns_ptrs.size());
+    thrust::transform_exclusive_scan(
+      thrust::host,
+      columns_ptrs.begin(),
+      columns_ptrs.end(),
+      offsets.begin(),
+      [](auto pcv) {
+        dictionary_column_view view(*pcv);
+        return offsets_pair{view.keys_size(), view.size()};
+      },
+      offsets_pair{0, 0},
+      [](auto lhs, auto rhs) {
+        return offsets_pair{lhs.first + rhs.first, lhs.second + rhs.second};
+      });
+    auto d_offsets = rmm::device_uvector<offsets_pair>(offsets.size(), stream);
+    CUDA_TRY(cudaMemcpyAsync(d_offsets.data(),
+                             offsets.data(),
+                             offsets.size() * sizeof(offsets_pair),
+                             cudaMemcpyHostToDevice,
+                             stream));
+    return d_offsets;
+  }
+
+ private:
+  std::vector<column_view const*> columns_ptrs;  ///< pointer version of input column_view vector
+};
+
+/**
+ * @brief Type-dispatch functor for remapping the old indices to new values based
+ * on the new key-set.
+ *
+ * The dispatch is based on the key type.
+ * The output column is the updated indices child for the new dictionary column.
+ */
+struct dispatch_compute_indices {
+  template <typename Element>
+  typename std::enable_if_t<cudf::is_relationally_comparable<Element, Element>(),
+                            std::unique_ptr<column>>
+  operator()(column_view const& all_keys,
+             column_view const& all_indices,
+             column_view const& new_keys,
+             offsets_pair const* d_offsets,
+             size_type const* d_map_to_keys,
+             cudaStream_t stream,
+             rmm::mr::device_memory_resource* mr)
+  {
+    auto keys_view     = column_device_view::create(all_keys, stream);
+    auto d_all_keys    = *keys_view;
+    auto indices_view  = column_device_view::create(all_indices, stream);
+    auto d_all_indices = *indices_view;
+
+    auto all_itr = thrust::make_transform_iterator(
```

Comment:
This might be more intuitive as a `permutation_iterator` over `keys_view->begin()`.